src/context/key_cursor.rs

```diff
@@ -0,0 +1,129 @@
+/// ```
+pub struct ScanKeyCursor {
+    key: RedisKey,
+    inner_cursor: *mut raw::RedisModuleScanCursor,
+}
+
+impl ScanKeyCursor {
+    /// Creates a new scan cursor for the given key.
+    pub fn new(key: RedisKey) -> Self {
+        let inner_cursor = unsafe { raw::RedisModule_ScanCursorCreate.unwrap()() };
+        Self { key, inner_cursor }
+    }
+
+    /// Restarts the cursor from the beginning.
+    pub fn restart(&self) {
+        unsafe { raw::RedisModule_ScanCursorRestart.unwrap()(self.inner_cursor) };
+    }
+
+    pub fn scan<F: FnMut(&RedisKey, &RedisString, &RedisString)>(&self, f: F) -> bool {
+        // the following is the callback definition. The callback may be called multiple times per `RedisModule_ScanKey` invocation.
+        use pimpl::scan_callback;
+
+        // Safety: The c-side initialized the function ptr and it is is never changed,
+        // i.e. after module initialization the function pointers stay valid till the end of the program.
+        let res = unsafe {
+            raw::RedisModule_ScanKey.unwrap()(
+                self.key.key_inner,
+                self.inner_cursor,
+                Some(scan_callback::<F>),
+                &f as *const F as *mut c_void,
+            )
+        };
+
+        res != 0
+    }
+
+    /// Implements a callback based for_each loop over all fields and values in the hash key, use that for optimal performance.
+    pub fn for_each<F: FnMut(&RedisKey, &RedisString, &RedisString)>(&self, mut f: F) {
+        while self.scan(&mut f) {
+            // do nothing, the callback does the work
+        }
+    }
+}
+
+impl Drop for ScanKeyCursor {
+    fn drop(&mut self) {
+        unsafe { raw::RedisModule_ScanCursorDestroy.unwrap()(self.inner_cursor) };
+    }
+}
+
+// the module contains the private implementation details of the cursor.
+mod pimpl {
+    use super::*;
+
+    /// The callback that is used by [`ScanKeyCursor::scan`] and [`ScanKeyCursor::for_each`] as argument to `RedisModule_ScanKey`.
+    ///
+    /// The `data` pointer is the closure given to [`ScanKeyCursor::for_each`] and the callback forwards
+    /// references to the key, field and value to that closure.
+    pub(super) unsafe extern "C" fn scan_callback<
+        F: FnMut(&RedisKey, &RedisString, &RedisString),
+    >(
+        key: *mut raw::RedisModuleKey,
+        field: *mut raw::RedisModuleString,
+        value: *mut raw::RedisModuleString,
+        data: *mut c_void,
+    ) {
+        let ctx = ptr::null_mut();
+        let key = RedisKey::from_raw_parts(ctx, key);
+
+        let field = RedisString::from_redis_module_string(ctx, field);
+        let value = RedisString::from_redis_module_string(ctx, value);
+
+        let callback = unsafe { &mut *(data.cast::<F>()) };
+        callback(&key, &field, &value);
+
+        // we're not the owner of field and value strings
+        field.take();
+        value.take();
+
+        key.take(); // we're not the owner of the key either
+    }
```

Comment:
just move this into the body of the `scan` method, so its private by default and we colocate the tricky implementation details